src/main/java/com/griddynamics/cloud/learning/dal/repositories/Repository.java

```diff
@@ -0,0 +1,78 @@
+import com.griddynamics.cloud.learning.dal.domain.DomainEntity;
+import com.griddynamics.cloud.learning.dal.dto.DomainEntitiesDTO;
+import com.griddynamics.cloud.learning.dal.dto.ArticleRequestParamsDTO;
+
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public abstract class Repository<T extends DomainEntity> {
+
+    private static final Logger logger = Logger.getLogger(Repository.class.getName());
+
+    protected abstract String getEntityTableName();
+
+    protected abstract Optional<T> convertToEntity(ResultSet resultSet);
+
+    public Optional<T> getById(final long id) {
+        String query = "SELECT * FROM " + getEntityTableName() + " WHERE id = " + id + ";";
+        return getEntityByParameter(query);
+    }
+
+    protected Optional<T> getEntityByParameter(final String query) {
+        try (Connection connection = DBConnectionManager.getConnection();
+             PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            resultSet.next();
+            return convertToEntity(resultSet);
+        } catch (SQLException e) {
+            final String msg = "Unable to retrieve required entity. Query \"" + query + "\" aborted.";
+            logger.log(Level.SEVERE, msg, e);
+            throw new AbortedQueryException(msg, e);
+        }
+    }
+
+    protected DomainEntitiesDTO<T> getEntitiesDTO(final String query, final ArticleRequestParamsDTO params) {
+
+        final int startRow = (params.getPageNumber() - 1) * params.getLimitPerPage();
+        final String paginatedQuery = query + " LIMIT " + params.getLimitPerPage() + " OFFSET " + startRow;
+
+        try (Connection connection = DBConnectionManager.getConnection();
+             PreparedStatement statement = connection.prepareStatement(query, ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.CONCUR_READ_ONLY);
+             PreparedStatement paginatedStatement = connection.prepareStatement(paginatedQuery)) {
+
+            ResultSet resultSet = statement.executeQuery();
+            int rowsCount = countRows(resultSet);
+
+            ResultSet paginatedResultSet = paginatedStatement.executeQuery();
+            List<T> entities = new ArrayList<>();
+            while (paginatedResultSet.next()) {
+                Optional<T> entity = convertToEntity(resultSet);
+                entity.ifPresent(entities::add);
+            }
+            return new DomainEntitiesDTO<T>(entities, rowsCount);
+        } catch (SQLException e) {
+            final String msg = "Unable to retrieve required entities. Query \"" + query + "\" aborted.";
+            logger.log(Level.SEVERE, msg, e);
+            throw new AbortedQueryException(msg, e);
+        }
+    }
+
+    private int countRows(final ResultSet resultSet) {
+        try {
+            resultSet.last();
```

Comment:
Here is a kind of inefficient operation, please check the source code.
I'd like to suggest the following approach:  instead of making almost same request you can simplify it, get rid from sorting, and already (from the request result) obtain a rows count. 
Example: Select count(*) from ... where caption ....